server/src/main/java/org/elasticsearch/cluster/features/NodeFeaturesFixListener.java

```diff
@@ -0,0 +1,215 @@
+        Scheduler scheduler,
+        Executor executor
+    ) {
+        this.taskQueue = taskQueue;
+        this.client = client;
+        this.scheduler = scheduler;
+        this.executor = executor;
+    }
+
+    class NodesFeaturesTask implements ClusterStateTaskListener {
+        private final Map<String, Set<String>> results;
+        private final int retryNum;
+
+        NodesFeaturesTask(Map<String, Set<String>> results, int retryNum) {
+            this.results = results;
+            this.retryNum = retryNum;
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            logger.error("Could not apply features for nodes {} to cluster state", results.keySet(), e);
+            scheduleRetry(results.keySet(), retryNum);
+        }
+
+        public Map<String, Set<String>> results() {
+            return results;
+        }
+    }
+
+    private static class NodesFeaturesUpdater implements ClusterStateTaskExecutor<NodesFeaturesTask> {
+        @Override
+        public ClusterState execute(BatchExecutionContext<NodesFeaturesTask> context) {
+            ClusterState.Builder builder = ClusterState.builder(context.initialState());
+            boolean modified = false;
+            for (var c : context.taskContexts()) {
+                for (var e : c.getTask().results().entrySet()) {
+                    var existingFeatures = builder.nodeFeatures();
+                    // double check there are still no features for the node
+                    if (existingFeatures.getOrDefault(e.getKey(), Set.of()).isEmpty()) {
+                        builder.putNodeFeatures(e.getKey(), e.getValue());
+                        modified = true;
+                    }
+                }
+                c.success(() -> {});
+            }
+            return modified ? builder.build() : context.initialState();
+        }
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.nodesDelta().masterNodeChanged() && event.localNodeMaster()) {
+            /*
+             * Execute this if we have just become master.
+             * Check if there are any nodes that understand the nodes features action, but don't have any features
+             * added to cluster state, and so do a query to get the node features separately.
+             * We can't use features to determine when this should happen, as the features are incorrect.
+             * So use transport version as a proxy instead.
+             */
+            ClusterFeatures nodeFeatures = event.state().clusterFeatures();
+            Map<String, CompatibilityVersions> nodeVersions = getCompatibilityVersions(event.state());
+            Set<String> queryNodes = event.state()
+                .nodes()
+                .stream()
+                .map(DiscoveryNode::getId)
+                .filter(n -> nodeVersions.get(n).transportVersion().onOrAfter(TransportVersions.NODE_FEATURES_QUERY_ACTION))
```

Comment:
This check is not actually reliable for upgrading from versions before 8.8.0, due to the transport version being inferred to start with. This would require this listener to run strictly after `TransportVersionsFixupListener` has completed, but I don't think it's possible to add a dependency in this way.

Reply:
We would need to integrate this with `TransportVersionsFixupListener` to enforce the ordering properly

Reply:
Using `Version` (option 2.) seems ok to me as well in this case.
Alternatively we could listen for the fix implemented by `NodeFeaturesFixListener` listener in addition to make sure the fix has been applied.